### PR TITLE
feat(#642): feed API

### LINKED ISSUE #642: Feed / insights panel
Source:  https://kmaevents.com/ 
"Привіт, я зробив агрегатор всіх могилянських сошок і подій, можеш, будь ласка, глянути і дати якісь поради/ідеї/зауваження?"

This implementation should consist of 2 stages:
1. PoC temporary static banner in the header 
2. Dynamic configurable feed screen

### PR DESCRIPTION
**Pull request by stasiaaleks**
## Summary
This PR implements Feed API. Main change:  `GET /api/v1/feed/` - a cursor-paginated stream that combines recent course reviews + admin-authored posts. Extra: wires API to the frontend.

**Backend**
- `FeedPost` model + admin. Manually created posts, can be, scheduled via `published_at`, can be pinned until unpinned.
- Pagination: keyset cursor (`FeedCursor`) over `(occurred_at, id)`. Offset can't page a merged two-table stream: neither table knows how many rows of the other come before it, and new reviews at the top shift everything down, so items get repeated or skipped.
- `FeedService` merges reviews and posts in memory, lean review query, comments-only.
-  One endpoint that returns both kinds of item, tagged with `kind`. 
- Caching: pages are cached and dropped when a rating or post changes - via a signal for posts, since admin edits never reach the service layer. Plus a Redis watermark that lets the first request past a scheduled post's publication time retire the cache - so a cache hit costs no DB query at all.

**Frontend**
- `useFeed` — `useInfiniteQuery` on the cursor + intersection-observer scroll.
- Review cards show an arrow comparing the score to the course average.
- Mock data deleted; affected tests rewritten against fixtures.

Resolves #642 

## Testing
Tests added, and feature branch was deployed to staging.

## Notes
- Hidden behind the existing `fe_feed` feature flag.

## Summary by CodeRabbit

* **New Features**
  * Added a mixed feed of course reviews and promotional posts with cursor-based pagination and pinned content.
  * Added admin tools for creating, scheduling, styling, and managing promotional posts.
  * Added automatic infinite scrolling, review comparisons with course averages, promotional CTAs, images, and accent styles.
  * Added loading, empty, error, retry, and unavailable states.
* **Bug Fixes**
  * Invalid feed cursors now return a clear validation error.
  * Promotional image uploads now accept only supported image formats.

### CODE REVIEW

File: src/webapp/src/features/feed/components/FeedStrip.tsx [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+ * Each item is a fixed-width card; promos keep their accent treatment inline
+ * so ads and review activity flow through the same carousel.
+ *
+ * Older entries are reached via the trailing "view all" tile and the header
+ * link (the `/feed` route) rather than by discovering the horizontal scroll,
+ * which isn't obvious on its own.
+ */
+export function FeedStrip() {
+	const { enabled, isReady } = useFeatureFlagState(FEED_FLAG);
+	const items = MOCK_FEED_ITEMS;
```
[the remark this answers is not in the document]
Comment (stasiaaleks) [reply src/webapp/src/features/feed/components/FeedStrip.tsx]: already present in recent commit

File: src/webapp/src/features/feed/feedFlags.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+/**
+ * Waffle flag that gates the homepage feed.
+ *
+ * Opt-in, following the `fe_`-prefixed convention: the feed renders only when
+ * the flag resolves on, so it ships dark and can be disabled on demand by
+ * turning the flag off (server-side, or via the `featureFlags` console helper
+ * in non-live builds).
+ */
+export const FEED_FLAG = "fe_feed";
```
[the remark this answers is not in the document]
Comment (stasiaaleks) [reply src/webapp/src/features/feed/feedFlags.ts]: already present in recent commit

File: src/backend/rating_app/models/feed_post.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    """
+    An admin-authored post in the feed.
+
+    Unlike `PromoBanner`, many posts are live at once — they are interleaved
+    with review activity by `published_at`. `is_active` is the kill switch,
+    `published_at` is the schedule: a post dated in the future stays hidden
+    until then.
+    """
+
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
```
[the remark this answers is not in the document]
Comment (stasiaaleks) [reply src/backend/rating_app/models/feed_post.py:22]: already present in recent commit

File: src/backend/rating_app/models/feed_post.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+from django.core.validators import FileExtensionValidator
+from django.db import models
+from django.utils import timezone
+
+from .choices import FeedPostAccent
+
+FEED_IMAGE_EXTENSIONS = ["svg", "png", "webp", "jpg", "jpeg"]
+
+
+class FeedPost(models.Model):
```
[the remark this answers is not in the document]
Comment (stasiaaleks) [reply src/backend/rating_app/models/feed_post.py:12]: already present in recent commit

File: src/backend/rating_app/serializers/feed.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    def get_items(self, page) -> list:
+        # PolymorphicProxySerializer documents the union but cannot serialize;
+        # dispatch on the DTO type instead.
+        return [self._serialize(item) for item in page.items]
+
+    def _serialize(self, item: FeedReviewItem | FeedPromoItem) -> dict:
+        serializer = (
+            FeedPromoItemSerializer if isinstance(item, FeedPromoItem) else FeedReviewItemSerializer
+        )
+        return serializer(item).data
```
[the remark this answers is not in the document]
Comment (stasiaaleks) [reply src/backend/rating_app/serializers/feed.py]: already present in recent commit

File: src/webapp/src/features/feed/components/FeedPromoItem.tsx [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+interface FeedPromoItemProps {
+	readonly item: FeedPromoItemType;
+	/** `banner` is wider/horizontal for top-of-page placement. */
+	readonly variant?: "card" | "banner";
+}
+
+export function FeedPromoItem({ item, variant = "card" }: FeedPromoItemProps) {
+	const accent = ACCENT_STYLES[item.accent ?? "brand"];
+	const label = item.label ?? "Оголошення";
+	const isBanner = variant === "banner";
```
[the remark this answers is not in the document]
Comment (stasiaaleks) [reply src/webapp/src/features/feed/components/FeedPromoItem.tsx:51]: already present in recent commit

File: src/backend/rating_app/services/feed_service.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+class FeedService:
+    def __init__(
+        self,
+        feed_post_repository: FeedPostRepository,
+        rating_repository: RatingRepository,
+    ):
+        self.feed_post_repository = feed_post_repository
+        self.rating_repository = rating_repository
+
+    @rcached(ttl=FEED_CACHE_TTL, versioned_by=FEED_NAMESPACE)
```
[the remark this answers is not in the document]
Comment (stasiaaleks) [reply src/backend/rating_app/services/feed_service.py]: fixed

[the remark this answers is not in the document]
Comment (stasiaaleks) [reply src/backend/rating_app/services/feed_service.py]: fixed

File: src/backend/rating_app/signals.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+from rateukma.caching.instances import redis_cache_manager
+from rateukma.caching.patterns import FEED_NAMESPACE
+from rating_app.models import FeedPost
+
+
+#! Not a domain-event observer: posts are authored in Django admin.
+# it writes straight to the ORM and never reaches the service layer.
+@receiver([post_save, post_delete], sender=FeedPost)
+def invalidate_feed_cache(sender, **kwargs) -> None:
+    redis_cache_manager().bump_version(FEED_NAMESPACE)
```
[the remark this answers is not in the document]
Comment (stasiaaleks) [reply src/backend/rating_app/signals.py]: fixed

File: src/backend/rateukma/caching/patterns.py [hunk 9]
```diff
@@ -7,6 +7,8 @@
 COURSES_LIST_NAMESPACE = "courses:list"
 ANALYTICS_LIST_NAMESPACE = "analytics:list"
 FILTER_OPTIONS_NAMESPACE = "courses:filter-options"
+FEED_NAMESPACE = "feed"
```
Comment (Fybex) [inline src/backend/rateukma/caching/patterns.py]:
```suggestion
FEED_NAMESPACE = "feed:list"
```

to match others?

File: src/backend/rating_app/caching/feed_watermark.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
+"""Marks when the next scheduled feed post goes live.
```
Comment (Fybex) [inline src/backend/rating_app/caching/feed_watermark.py]: that's really cool idea, tho I'm not sure whether we need it, 36 line more or less, don't matter that much, but maybe it adds unnecessary complexity

do we have case where it's important to see update at 12:00, not in 12:01

alternative could be to build ttl dynamically maybe, but it may be also overkill

sth like:
```py
ttl = FEED_CACHE_TLL
next_at = repo.get_next_future_publication_time()
if next_at:
    ttl = min of both... to reset before 12:00
```

File: src/backend/rating_app/pagination/cursor.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+from datetime import UTC, datetime
+from uuid import UUID
+
+from django.db.models import Q
+
+from rating_app.exception.feed_exceptions import InvalidCursorError
+
+
+@dataclass(frozen=True, slots=True)
+class FeedCursor:
```
Comment (Fybex) [inline src/backend/rating_app/pagination/cursor.py:13]: was there option of separate feed events table? do you think we will have mroe sources than rating and feed post in future or not? 

`src/backend/rating_app/pagination/cursor.py` and if it's placed in quite generic place, can we avoid using "Feed" here maybe or name `feed_cursor.py` (nit-pick)?

File: src/backend/rating_app/services/feed_service.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+FEED_CACHE_TTL = 60
+
+FeedItem = FeedReviewItem | FeedPromoItem
+
+
+class FeedService:
+    def __init__(
+        self,
+        feed_post_repository: FeedPostRepository,
```
Comment (Fybex) [inline src/backend/rating_app/services/feed_service.py:24]: super nitpick: maybe take list of protocols?
```py
class FeedSource(Protocol):
    def get_page(self, cursor: FeedCursor | None, limit: int) -> list[FeedItem]: ...
```

File: src/backend/rating_app/services/feed_service.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+        if is_watermark_due(cache_manager):
+            cache_manager.bump_version(FEED_NAMESPACE)
+            store_feed_watermark(
+                cache_manager, self.feed_post_repository.get_next_future_publication_time()
+            )
+
+        return [FEED_NAMESPACE]
+
+    # `_cache_namespaces` is passed as a function object, so it has to be defined
```
Comment (Fybex) [inline src/backend/rating_app/services/feed_service.py:45]: nitpick: a lot of comments, not sure we need all of them, create cogntive load

File: src/webapp/src/features/feed/feedTypes.ts [hunk 14]
```diff
@@ -35,7 +34,7 @@ export interface FeedReviewItem {
  * Accent controls the promo card's color treatment so admins can visually
  * prioritise announcements without touching code.
  */
-export type FeedPromoAccent = "brand" | "info" | "warning";
+export type FeedPromoAccent = "BRAND" | "INFO" | "WARNING";
```
Comment (Fybex) [inline src/webapp/src/features/feed/feedTypes.ts]: don't we have this in generated types?

File: src/backend/rateukma/caching/patterns.py (hunk 9 above)
Comment (stasiaaleks) [reply src/backend/rateukma/caching/patterns.py]: ++, will adjust

File: src/backend/rating_app/pagination/cursor.py (hunk 11 above)
Comment (stasiaaleks) [reply src/backend/rating_app/pagination/cursor.py:13]: ++, will fix the nit

File: src/backend/rating_app/services/feed_service.py (hunk 12 above)
Comment (stasiaaleks) [reply src/backend/rating_app/services/feed_service.py:24]: will do in the next generic cursor refactoring PR ++

File: src/backend/rating_app/signals.py [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+from rateukma.caching.patterns import FEED_NAMESPACE
+from rating_app.ioc_container.services import feed_service
+from rating_app.models import FeedPost
+
+
+#! Not a domain-event observer: posts are authored in Django admin.
+# it writes straight to the ORM and never reaches the service layer.
+@receiver([post_save, post_delete], sender=FeedPost)
+def invalidate_feed_cache(sender, **kwargs) -> None:
+    redis_cache_manager().bump_version(FEED_NAMESPACE)
```
Comment (Fybex) [inline src/backend/rating_app/signals.py]: **Blocker.** Admin saves inside a transaction, and `post_save` runs before COMMIT. So the cache is bumped while the new post is still invisible to other requests. A feed request in that gap rebuilds the page without the post and caches it under the new version; the post then stays hidden for 60 s.

Run the bump after commit:
```python
@receiver([post_save, post_delete], sender=FeedPost)
def invalidate_feed_cache(sender, using, **kwargs) -> None:
    def bump():
        redis_cache_manager().bump_version(FEED_NAMESPACE)
        feed_service().refresh_next_publish_marker()
    transaction.on_commit(bump, using=using)
```
Test that catches it (needs `django_db(transaction=True)`):
```python
def test_post_saved_inside_a_transaction_is_not_cached_away(service):
    service.get_feed_page(cursor=None, limit=10)
    with transaction.atomic():
        post = FeedPostFactory(published_at=timezone.now() - timedelta(minutes=1))
        with patch.object(FeedPostRepository, "get_page", return_value=[]):  # reader that can't see the row yet
            service.get_feed_page(cursor=None, limit=10)
    assert str(post.id) in _ids(service.get_feed_page(cursor=None, limit=10))
```

_Published by Claude Code on Andrii's behalf._

File: src/backend/rating_app/migrations/0033_feedpost.py [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+import django.core.validators
+import django.utils.timezone
+import uuid
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("rating_app", "0032_deprecate_rating_instructor"),
```
Comment (Fybex) [inline src/backend/rating_app/migrations/0033_feedpost.py:11]: **Blocker.** `main` now has `0033_course_title_trgm_idx` (#668), so this `0033` conflicts. Squash `0034` into this file, rename to `0034_feedpost`, depend on `0033_course_title_trgm_idx`. CI now runs `makemigrations --check`.

_Published by Claude Code on Andrii's behalf._

File: src/webapp/src/features/feed/components/FeedPromoItem.tsx [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+};
+
+interface FeedPromoItemProps {
+	readonly item: FeedPromoItemType;
+	/** `banner` is wider/horizontal for top-of-page placement. */
+	readonly variant?: "card" | "banner";
+}
+
+export function FeedPromoItem({ item, variant = "card" }: FeedPromoItemProps) {
+	const accent = ACCENT_STYLES[item.accent ?? "BRAND"];
```
Comment (Fybex) [inline src/webapp/src/features/feed/components/FeedPromoItem.tsx]: `??` only handles `undefined`. If the backend sends an accent this bundle doesn't know, `ACCENT_STYLES[...]` is `undefined` and `.container` throws.
```ts
const accent = ACCENT_STYLES[item.accent ?? "BRAND"] ?? ACCENT_STYLES.BRAND;
```

_Published by Claude Code on Andrii's behalf._

File: src/backend/rating_app/repositories/feed_post_repository.py [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+from rating_app.repositories.to_domain_mappers import FeedPostMapper
+
+
+class FeedPostRepository:
+    """Read-only access to feed posts. Rows are authored in Django admin."""
+
+    def __init__(self, mapper: FeedPostMapper) -> None:
+        self._mapper = mapper
+
+    def get_pinned(self) -> list[FeedPromoItemDTO]:
```
Comment (Fybex) [inline src/backend/rating_app/repositories/feed_post_repository.py:20]: No limit on pinned posts: 10 pinned = 18 cards on the homepage strip. `posts[:MAX_PINNED]` with `MAX_PINNED = 3` and a note in the `pinned` help text.

_Published by Claude Code on Andrii's behalf._

File: src/webapp/src/routes/feed.tsx [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+					<h1 className="flex items-center gap-2 text-2xl font-bold tracking-tight">
+						<Newspaper className="size-6 text-muted-foreground" />
+						Стрічка оновлень
+					</h1>
+					<p className="text-muted-foreground">
+						Останні відгуки та оголошення Могилянки
+					</p>
+				</header>
+
+				{!isReady ? null : !enabled ? (
```
Comment (Fybex) [inline src/webapp/src/routes/feed.tsx]: Plain `<p>` for loading / error / empty. Follow `my-ratings.tsx`:
```tsx
if (isLoading) return <MyRatingsSkeleton />;
if (isError) return <MyRatingsErrorState onRetry={refetch} />;
if (items.length === 0) return <MyRatingsEmptyState />;
```
Needs `refetch` returned from `useFeed` and `feed.*` ids in `lib/test-ids.ts` so #661 can find the states.

_Published by Claude Code on Andrii's behalf._

File: src/webapp/src/features/feed/feedTestFixtures.ts [hunk 20]
```diff
@@ -0,0 +1,50 @@
+import type { FeedItem } from "./feedTypes";
```
Comment (Fybex) [inline src/webapp/src/features/feed/feedTestFixtures.ts]: We don't have fixture files; test data lives in `test-utils/factories.ts` as `createMockX(overrides)`. Move these there (`createMockFeedPromoItem`, `createMockFeedReviewItem`, `createMockFeedState`) and delete this file. `useFeed.ts` also has no test yet (cursor threading, `next_cursor: null`, error).

_Published by Claude Code on Andrii's behalf._

File: src/webapp/src/features/feed/hooks/useFeed.ts [hunk 21]
```diff
@@ -0,0 +1,142 @@
+import type { RefObject } from "react";
+import { useEffect, useMemo, useRef } from "react";
+
+import { useInfiniteQuery } from "@tanstack/react-query";
+
+import type { FeedItem as ApiFeedItem } from "@/lib/api/generated";
+import { feedList, getFeedListQueryKey } from "@/lib/api/generated";
+import type { FeedItem, FeedPromoItem, FeedReviewItem } from "../feedTypes";
```
Comment (Fybex) [inline src/webapp/src/features/feed/hooks/useFeed.ts:6]: `orval.config.ts` already generates infinite hooks (see `instructors_list`). Add
```ts
feed_list: { query: { useInfinite: true, useInfiniteQueryParam: "cursor" } },
```
and use `useFeedListInfinite` like `useInfiniteInstructors.ts` does.

_Published by Claude Code on Andrii's behalf._

File: src/backend/rating_app/models/feed_post.py [hunk 22]
```diff
@@ -0,0 +1,85 @@
+import uuid
+
+from django.core.validators import FileExtensionValidator
+from django.db import models
+from django.utils import timezone
+
+from .choices import FeedPostAccent
+
+FEED_IMAGE_EXTENSIONS = ["svg", "png", "webp", "jpg", "jpeg"]
```
Comment (Fybex) [inline src/backend/rating_app/models/feed_post.py]: Drop `svg`: it is served from the API origin and can run scripts there.

_Published by Claude Code on Andrii's behalf._

**Review by Fybex — CHANGES_REQUESTED** [review]
Re-review at `09baf794`. Thanks for the `feed:list` rename and the `AccentEnum` alias.

Blocking: `signals.py` bump before COMMIT, migration renumbering (`main` now has `0033_course_title_trgm_idx`), accent runtime fallback.

Non-blocking: pinned cap, route states, fixtures -> factories, orval infinite override, `svg`.

Rebase on `main`: it already has `fast-uri: 3.1.7`, and please leave `ioc_container/services.py`, `views/responses.py`, `views/__init__.py` untouched (CRLF->LF rewrites, ~10 real lines) and keep the `PromoBannerAdmin` description.

Please resolve review threads only after the fixing commit is pushed.

_Published by Claude Code on Andrii's behalf._

**Review by Fybex — APPROVED** [review]
